Update foreign key handling to be smarter around when to skip foreign key emulation

arthurschreiber opened the pull request:
## Description

This improves the foreign key implementation by being more clever about when we can push the entire DML statement down to MySQL and when vtgate has to emulate the foreign key handling itself.

For every DML statement that this planner sees — `UPDATE`, `DELETE`, `INSERT`, `REPLACE`, and `INSERT ... ON DUPLICATE KEY UPDATE` — vtgate can push the statement straight to MySQL when **every** involved foreign key satisfies all of:

* The child foreign key's relevant referential action is `RESTRICT`, `NO ACTION`, or `DEFAULT` — i.e. an action MySQL can enforce in-place without cascading. `UPDATE` (and the `ON DUPLICATE KEY UPDATE` half of `INSERT`) looks at `ON UPDATE`; `DELETE` and `REPLACE` look at `ON DELETE`.
* The foreign key does not cross a keyspace boundary.
* The foreign key is shard-local (parent and child share the same vindex / land on the same shard) for sharded keyspaces.

The decision is made by a single helper, `SemTable.RequiresForeignKeyEmulation(getAction)`, that is called from the `UPDATE`, `DELETE`, and `INSERT` planbuilders. When the helper reports that no emulation is required, all involved foreign keys are cleared from the semantic table and the query follows the regular (non-FK-aware) plan path.

### Tests

* Unit tests in `go/vt/vtgate/semantics/semantic_table_test.go` cover the decision logic for parent/child FKs in same/different keyspaces, sharded vs. unsharded placement, every referential action (`RESTRICT`, `CASCADE`, `SET NULL`, `SET DEFAULT`), multi-expression `UPDATE`, multi-target `UPDATE`, `DELETE`, `INSERT`, `REPLACE`, and `INSERT ... ON DUPLICATE KEY UPDATE`.
* `TestFkFuzzTest` now runs against the shard-scoped keyspace in addition to the unsharded one, exercising the RESTRICT push-down path and the CASCADE / `SET NULL` emulation path on real sharded data. The fuzzer also generates `INSERT ... ON DUPLICATE KEY UPDATE` queries.

## Related Issue(s)

Closes #19911

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

## Deployment Notes

arthurschreiber committed:
Update foreign key checks to be smarter around when to skip (or not skip) foreign key emulation.

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Fix unit test cases.

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Merge branch 'main' of https://github.com/vitessio/vitess into arthur/smarter-fk-handling

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Change foreign key emulation check.

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Undo some of the changes.

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Revert incorrect logic removal.

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Simplify foreign key check.

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
cleanups

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Merge branch 'main' of https://github.com/vitessio/vitess into arthur/smarter-fk-handling

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Add unit tests.

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

harshit-gangal reviewed (commented):
The changes looks good.
We should add some self referential fuzzer testing scenario in e2e test.

arthurschreiber committed:
Merge branch 'main' of https://github.com/vitessio/vitess into arthur/smarter-fk-handling

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

mattlord reviewed (commented):
Seems reasonable to me, although I'm not sure why we can't passthrough regardless of the action type.

@arthurschreiber please create a feature/enhancement request and link it to this PR. It doesn't have to be much, but describing what the desired improvement is and the use case (you largely did this in the PR description).

Thanks!

arthurschreiber committed:
Merge remote-tracking branch 'upstream/main' into arthur/smarter-fk-handling

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Add self-referential CASCADE table to foreign key fuzzer

Adds fk_t21 with a self-referential foreign key that cascades on
delete, complementing fk_t20's RESTRICT self-reference. InnoDB does
not activate non-RESTRICT ON UPDATE actions on self-referential
tables, so only the ON DELETE action is varied.

Extracts the self-referential table detection into an isSelfRefFkTable
helper so new self-referential tables can be wired in uniformly.

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Guard RequiresForeignKeyEmulation against ambiguous column deps

When an update expression's column resolves to zero or more than one
table via its recursive dependencies, TableOffset() returns -1 and
indexing st.Tables with that offset panics. This can happen when the
column comes from a derived table whose projection spans multiple base
tables.

Conservatively require emulation in that case so vtgate re-validates
all involved foreign keys rather than pushing to MySQL.

Reported in PR #18618 review.

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Revert "Add self-referential CASCADE table to foreign key fuzzer"

This reverts commit fa18ed1a58f77cf06284ecb7c3bcd92fa74a05e2.

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Address PR review feedback on foreign key emulation

* Clarify RequiresForeignKeyEmulation doc comment to reference
  ReferenceAction.IsRestrict instead of listing individual actions,
  covering RESTRICT, NO ACTION, and DefaultAction uniformly.
* Replace the manual delete-in-range in ClearForeignKeys with the
  built-in clear().
* Fix malformed comment in planbuilder/update.go (//If -> // If).
* Expand TestRequiresForeignKeyEmulation with SET NULL, SET DEFAULT,
  multi-expression UPDATE, and multi-target UPDATE cases.
* Generalize RequiresForeignKeyEmulation to accept a ChildFKInfo
  action getter (matching RemoveNonRequiredForeignKeys) and iterate
  the involved foreign key maps directly. This lets DELETE and INSERT
  short-circuit FK-aware planning the same way UPDATE does. Add the
  corresponding ClearForeignKeys call sites in delete.go and insert.go
  and cover DELETE, INSERT, REPLACE, and INSERT ... ON DUPLICATE KEY
  UPDATE in the tests.

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Generate INSERT ON DUPLICATE KEY UPDATE in the foreign key fuzzer

Extends the INSERT query generator to sometimes emit an ON DUPLICATE
KEY UPDATE clause, covering the plain SQL, prepared statement, and
prepared packet paths. Only INSERT carries the clause; REPLACE does
not support it.

This exercises the INSERT-with-ON-DUPLICATE push-down path that the
planbuilder handles via UpdateAction.

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Run foreign key fuzzer against the shard-scoped keyspace

Adds a shardScoped flag to the fuzzer that restricts generation to
query forms that do not mutate a vindex column (INSERTs without ON
DUPLICATE KEY UPDATE, and DELETEs). In the shard-scoped vschema the
FK parent/child columns double as the shard vindex, so UPDATE and
REPLACE against them are unsupported by Vitess and would produce
Vitess-vs-MySQL error mismatches.

With this flag wired up, TestFkFuzzTest now iterates shardScopedKs in
addition to unshardedKs, exercising the shard-scoped RESTRICT
push-down path and the shard-scoped CASCADE emulation path against
real data. Replica FK constraints on shardScopedKs are also cleaned
up up-front like the other keyspaces.

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
gofumpt: reformat RequiresForeignKeyEmulation doc comment bullets

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber committed:
Use ON DELETE action getter for REPLACE foreign key push-down decision

REPLACE performs a DELETE followed by an INSERT; the child foreign
keys it fires are ON DELETE actions, not ON UPDATE. Previously the
INSERT planbuilder always passed UpdateAction to both
RequiresForeignKeyEmulation and RemoveNonRequiredForeignKeys, so a
table with ON DELETE CASCADE and ON UPDATE RESTRICT would be
classified as push-down safe for REPLACE even though the delete half
would cascade.

Pick the action getter from the insert action: ReplaceAct gets
DeleteAction, everything else (plain INSERT, INSERT ... ON DUPLICATE
KEY UPDATE) keeps UpdateAction.

Also replace the REPLACE 'cascading' unit test with two sharper
cases: ON DELETE CASCADE (requires emulation) and ON UPDATE CASCADE
(does not, since REPLACE never fires ON UPDATE).

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

arthurschreiber commented on go/vt/vtgate/semantics/semantic_table_test.go:
```diff
@@ -870,6 +870,671 @@ func TestIsFkDependentColumnUpdated(t *testing.T) {
+					"t1": {
+						Name:     sqlparser.NewIdentifierCS("t1"),
+						Keyspace: unshardedKeyspace,
+					},
+				},
+			},
+			requiresEmulation: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
```
This was fixed in https://github.com/vitessio/vitess/pull/18618/commits/d6fba7a968ba83f47a5e67f9ee50c614783fed71.

arthurschreiber commented on go/vt/vtgate/semantics/semantic_table_test.go:
```diff
@@ -870,6 +870,647 @@ func TestIsFkDependentColumnUpdated(t *testing.T) {
+func actionForStmt(stmt sqlparser.Statement) func(vindexes.ChildFKInfo) sqlparser.ReferenceAction {
+	switch stmt.(type) {
+	case *sqlparser.Delete:
+		return vindexes.DeleteAction
```
This was fixed in https://github.com/vitessio/vitess/pull/18618/commits/d6fba7a968ba83f47a5e67f9ee50c614783fed71.

arthurschreiber commented on go/vt/vtgate/planbuilder/insert.go:
```diff
@@ -44,10 +44,14 @@ func gen4InsertStmtPlanner(version querypb.ExecuteOptions_PlannerVersion, insStm
+		if err != nil {
+			return nil, err
+		}
 	}
```
This was fixed in https://github.com/vitessio/vitess/pull/18618/commits/d6fba7a968ba83f47a5e67f9ee50c614783fed71.

arthurschreiber committed:
Fix shard-scoped FK fuzzer table selection

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

mattlord reviewed (approved):
The only non-blocking issue I see is that the new shard-scoped fuzzer run still forces `updateShare = 0` and disables `REPLACE` / `INSERT` `... ON DUPLICATE KEY UPDATE` (ODKU), and it also excludes the self-referential `fk_t20` case. So the new sharded real-data coverage doesn’t actually exercise the highest-risk planner paths this PR changes; those `UPDATE` / `REPLACE` / `ODKU` decisions are still unit-test-only. It would be safer if we had at least one targeted sharded e2e test for a shard-local `RESTRICT` update, and ideally one `REPLACE` or `ODKU` case too.

arthurschreiber committed:
Fix foreign key fuzz test generation

Signed-off-by: Arthur Schreiber <arthur@planetscale.com>

timvaillancourt commented:
Hey @arthurschreiber 👋 heads up, `vtgate_foreignkey_stress` is failing on the head commit:
https://github.com/vitessio/vitess/actions/runs/24928774704

`go/test/endtoend/vtgate/foreignkey` dies with `panic: test timed out after 10m0s`, and the
`TestFkQueries` "failures" are just casualties of the timeout (`-1.00s`)

timvaillancourt reviewed (commented):
I gave this another pass with a few AI reviewers; all 3 x traced the skip decision independently and agree it's correct (`IsRestrict()` is the exact complement of the keep-set in `RemoveNonRequiredForeignKeys`, and the previously-`VT12001` push-downs check out), so this is close to a stamp from me. A few things first:

1. `RequiresForeignKeyEmulation` can panic on the new path: `isShardScoped` indexes `ColumnVindexes[0]` on both tables unguarded (`semantic_table.go:325`), and a managed FK in a sharded keyspace where one side has no primary vindex now reaches it; the old path returned early via `VerifyAllFKs` for the non-literal update case. Treating a missing primary vindex as "requires emulation" seems like the safe guard
2. The `VT12001` error-to-success change and the `FKVerify`/`ForeignKey` to `Passthrough` plan flips are user-visible, so I believe this needs a `changelog/25.0` entry
3. The shard-scoped fuzzer runs with `updateShare = 0` and `allowReplace`/`allowOnDup` disabled, so the riskiest new path (shard-scoped `UPDATE` of an FK column pushed down with checks ON) never runs end-to-end against the comparison DB; related to the earlier comment that the PR description mentions `INSERT ... ON DUPLICATE KEY UPDATE` fuzzing that's currently disabled
4. And the `vtgate_foreignkey_stress` timeout from my earlier comment is still there, probably just needs the test timeout bumped now that the fuzzer runs 2 x keyspaces

Happy to re-review once these are in
